httpd-client/index.ts

```diff
@@ -0,0 +1,136 @@
+import type { BaseFetch } from "./lib/fetcher";
+import type { RequestOptions, Method } from "./lib/fetcher";
+import type { ZodSchema } from "zod";
+
+import type { Blob, Project, Remote, Tree } from "./lib/project";
+import type {
+  Commit,
+  CommitHeader,
+  Diff,
+  DiffAddedDeletedModifiedChangeset,
+  HunkLine,
+} from "./lib/project/commit";
+
+import { array, literal, number, strictObject, string, union } from "zod";
+
+import * as project from "./lib/project";
+import * as session from "./lib/session";
+import { Fetcher } from "./lib/fetcher";
+
+export type {
+  Blob,
+  Commit,
+  CommitHeader,
+  Diff,
+  DiffAddedDeletedModifiedChangeset,
+  HunkLine,
+  Project,
+  Remote,
+  Tree,
+};
+
+export interface BaseUrl {
+  host: string;
+  port: number;
+  scheme: string;
+}
+
+export interface Root {
+  message: string;
+  service: string;
+  version: string;
+  node: { id: string };
+  path: string;
+  links: { href: string; rel: string; type: Method }[];
+}
+
+const rootSchema: ZodSchema<Root> = strictObject({
+  message: string(),
+  service: string(),
+  version: string(),
+  node: strictObject({ id: string() }),
+  path: string(),
+  links: array(
+    strictObject({
+      href: string(),
+      rel: string(),
+      type: union([
+        literal("GET"),
+        literal("POST"),
+        literal("PUT"),
+        literal("DELETE"),
+      ]),
+    }),
+  ),
+});
+
+export interface NodeStats {
+  projects: { count: number };
+  users: { count: number };
+}
+
+const nodeStatsSchema: ZodSchema<NodeStats> = strictObject({
+  projects: strictObject({ count: number() }),
+  users: strictObject({ count: number() }),
+});
+
+export interface Node {
+  id: string;
+}
+
+const nodeSchema: ZodSchema<Node> = strictObject({
+  id: string(),
+});
+
+export class HttpdClient {
+  private fetcher: Fetcher;
+  #baseUrl: string;
+
+  public project: project.Client;
+  public session: session.Client;
+
+  public constructor(baseUrl: string | BaseUrl, fetch?: BaseFetch) {
```

Comment:
You can simplify the API if you require `baseUrl` to be of type `BaseUrl`. From what I can tell the string type is only really used as a shortcut and it’s easy to implement the shortcut at the call site. I think in general it’s desirable to keep a library like this as straightforward as possible.

Reply:
bbcd2ab5a38caba1408d7f864b4fe853456b1199